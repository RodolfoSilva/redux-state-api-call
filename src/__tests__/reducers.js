import loadingReducer from '../reducers/loading';
import errorReducer from '../reducers/error';

describe('reducer', () => {
  describe('loading', () => {
    it('should return the initial state', () => {
      expect(loadingReducer(undefined, {})).toMatchSnapshot();
    });

    test.each([
      [
        { type: 'ADD_TODO_REQUEST' },
        undefined,
        { ADD_TODO: expect.any(Number) },
      ],
      [
        { type: 'ADD_TODO_FAILURE' },
        undefined,
        { ADD_TODO: expect.any(Number) },
      ],
      [
        { type: 'ADD_TODO_SUCCESS' },
        undefined,
        { ADD_TODO: expect.any(Number) },
      ],
    ])(
      'should handle actions ended with _(REQUEST|FAILURE|SUCCESS)',
      (action, state, expected) => {
        expect(loadingReducer(state, action)).toEqual(expected);
        expect(loadingReducer(state, action)).toMatchSnapshot();
      },
    );

    test.each([
      [{ type: 'ADD_TODO_FAILURE' }, undefined, { ADD_TODO: -1 }],
      [{ type: 'ADD_TODO_SUCCESS' }, undefined, { ADD_TODO: -1 }],
      [{ type: 'ADD_TODO_FAILURE' }, { ADD_TODO: 1 }, { ADD_TODO: 0 }],
      [{ type: 'ADD_TODO_FAILURE' }, { ADD_TODO: 3 }, { ADD_TODO: 2 }],
      [{ type: 'ADD_TODO_SUCCESS' }, { ADD_TODO: 1 }, { ADD_TODO: 0 }],
      [{ type: 'ADD_TODO_SUCCESS' }, { ADD_TODO: 3 }, { ADD_TODO: 2 }],
    ])(
      'should handle actions and decrement with _(FAILURE|SUCCESS)',
      (action, state, expected) => {
        expect(loadingReducer(state, action)).toEqual(expected);
        expect(loadingReducer(state, action)).toMatchSnapshot();
      },
    );

    test.each([
      [{ type: 'ADD_TODO_REQUEST' }, undefined, { ADD_TODO: 1 }],
      [{ type: 'ADD_TODO_REQUEST' }, { ADD_TODO: 1 }, { ADD_TODO: 2 }],
    ])(
      'should handle actions and increment with _REQUEST',
      (action, state, expected) => {
        expect(loadingReducer(state, action)).toEqual(expected);
        expect(loadingReducer(state, action)).toMatchSnapshot();
      },
    );
  });

  describe('error', () => {
    it('should return the initial state', () => {
      expect(errorReducer(undefined, {})).toMatchSnapshot();
    });

    test.each([
      [{ type: 'helloWorld' }, undefined, {}],
      [{ type: 'ADD_TODO_FULFILLED' }, undefined, {}],
      [{ type: 'ADD_TODO_REJECTED' }, undefined, {}],
      [{ type: 'ADD_TODO_PENDING' }, undefined, {}],
    ])(
      'should ignore actions not ended with _(REQUEST|SUCCESS|FAILURE)',
      (action, state, expected) => {
        expect(errorReducer(state, action)).toEqual(expected);
        expect(errorReducer(state, action)).toMatchSnapshot();
      },
    );

    test.each([
      [{ type: 'ADD_TODO_REQUEST' }, undefined, { ADD_TODO: null }],
      [{ type: 'ADD_TODO_SUCCESS' }, undefined, { ADD_TODO: null }],
      [{ type: 'ADD_TODO_FAILURE' }, undefined, { ADD_TODO: null }],
    ])(
      'should handle actions ended with _(REQUEST|SUCCESS|FAILURE)',
      (action, state, expected) => {
        expect(errorReducer(state, action)).toEqual(expected);
        expect(errorReducer(state, action)).toMatchSnapshot();
      },
    );

    test.each([
      [
        { type: 'ADD_TODO_REQUEST' },
        { ADD_TODO: new Error('Unexpected error') },
        { ADD_TODO: null },
      ],
      [
        { type: 'ADD_TODO_SUCCESS' },
        { ADD_TODO: new Error('Unexpected error') },
        { ADD_TODO: null },
      ],
    ])(
      'should handle actions ended with _(REQUEST|SUCCESS)',
      (action, state, expected) => {
        expect(errorReducer(state, action)).toEqual(expected);
        expect(errorReducer(state, action)).toMatchSnapshot();
      },
    );

    test.each([
      [
        {
          type: 'ADD_TODO_FAILURE',
          error: true,
          payload: new Error('Unexpected error'),
        },
        undefined,
        { ADD_TODO: new Error('Unexpected error') },
      ],
    ])(
      'should handle actions ended with _FAILURE',
      (action, state, expected) => {
        expect(errorReducer(state, action)).toEqual(expected);
        expect(errorReducer(state, action)).toMatchSnapshot();
      },
    );
  });
});
