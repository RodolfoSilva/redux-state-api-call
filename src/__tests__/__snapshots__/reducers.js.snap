// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`reducer error should handle actions ended with _(REQUEST|SUCCESS) 1`] = `
Object {
  "ADD_TODO": null,
}
`;

exports[`reducer error should handle actions ended with _(REQUEST|SUCCESS) 2`] = `
Object {
  "ADD_TODO": null,
}
`;

exports[`reducer error should handle actions ended with _(REQUEST|SUCCESS|FAILURE) 1`] = `
Object {
  "ADD_TODO": null,
}
`;

exports[`reducer error should handle actions ended with _(REQUEST|SUCCESS|FAILURE) 2`] = `
Object {
  "ADD_TODO": null,
}
`;

exports[`reducer error should handle actions ended with _(REQUEST|SUCCESS|FAILURE) 3`] = `
Object {
  "ADD_TODO": null,
}
`;

exports[`reducer error should handle actions ended with _FAILURE 1`] = `
Object {
  "ADD_TODO": [Error: Unexpected error],
}
`;

exports[`reducer error should ignore actions not ended with _(REQUEST|SUCCESS|FAILURE) 1`] = `Object {}`;

exports[`reducer error should ignore actions not ended with _(REQUEST|SUCCESS|FAILURE) 2`] = `Object {}`;

exports[`reducer error should ignore actions not ended with _(REQUEST|SUCCESS|FAILURE) 3`] = `Object {}`;

exports[`reducer error should ignore actions not ended with _(REQUEST|SUCCESS|FAILURE) 4`] = `Object {}`;

exports[`reducer error should return the initial state 1`] = `Object {}`;

exports[`reducer loading should handle actions and decrement with _(FAILURE|SUCCESS) 1`] = `
Object {
  "ADD_TODO": -1,
}
`;

exports[`reducer loading should handle actions and decrement with _(FAILURE|SUCCESS) 2`] = `
Object {
  "ADD_TODO": -1,
}
`;

exports[`reducer loading should handle actions and decrement with _(FAILURE|SUCCESS) 3`] = `
Object {
  "ADD_TODO": 0,
}
`;

exports[`reducer loading should handle actions and decrement with _(FAILURE|SUCCESS) 4`] = `
Object {
  "ADD_TODO": 2,
}
`;

exports[`reducer loading should handle actions and decrement with _(FAILURE|SUCCESS) 5`] = `
Object {
  "ADD_TODO": 0,
}
`;

exports[`reducer loading should handle actions and decrement with _(FAILURE|SUCCESS) 6`] = `
Object {
  "ADD_TODO": 2,
}
`;

exports[`reducer loading should handle actions and increment with _REQUEST 1`] = `
Object {
  "ADD_TODO": 1,
}
`;

exports[`reducer loading should handle actions and increment with _REQUEST 2`] = `
Object {
  "ADD_TODO": 2,
}
`;

exports[`reducer loading should handle actions ended with _(REQUEST|FAILURE|SUCCESS) 1`] = `
Object {
  "ADD_TODO": 1,
}
`;

exports[`reducer loading should handle actions ended with _(REQUEST|FAILURE|SUCCESS) 2`] = `
Object {
  "ADD_TODO": -1,
}
`;

exports[`reducer loading should handle actions ended with _(REQUEST|FAILURE|SUCCESS) 3`] = `
Object {
  "ADD_TODO": -1,
}
`;

exports[`reducer loading should return the initial state 1`] = `Object {}`;
